src/test/java/org/opensearch/neuralsearch/ml/MLCommonsClientAccessorTests.java

```diff
@@ -161,6 +196,62 @@ public void testInferenceSentences_whenNotConnectionException_thenNoRetry() {
         Mockito.verify(resultListener).onFailure(illegalStateException);
     }
 
+    public void testInferenceSentences_whenModelAsymmetric_thenSuccess() {
+        final List<Float> vector = new ArrayList<>(List.of(TestCommonConstants.PREDICT_VECTOR_ARRAY));
+        Mockito.doAnswer(invocation -> {
+            final ActionListener<MLOutput> actionListener = invocation.getArgument(2);
+            actionListener.onResponse(createModelTensorOutput(TestCommonConstants.PREDICT_VECTOR_ARRAY));
+            return null;
+        }).when(client).predict(Mockito.eq(TestCommonConstants.MODEL_ID), Mockito.isA(MLInput.class), Mockito.isA(ActionListener.class));
+        setupMocksForTextEmbeddingModelAsymmetryCheck(true);
+
+        accessor.inferenceSentence(
+            TestCommonConstants.MODEL_ID,
+            TestCommonConstants.SENTENCES_LIST.get(0),
+            AsymmetricTextEmbeddingParameters.builder().embeddingContentType(EmbeddingContentType.PASSAGE).build(),
+            singleSentenceResultListener
+        );
+
+        Mockito.verify(client)
+            .predict(
+                Mockito.eq(TestCommonConstants.MODEL_ID),
+                Mockito.argThat((MLInput input) -> input.getParameters() != null),
+                Mockito.isA(ActionListener.class)
+            );
+        Mockito.verify(singleSentenceResultListener).onResponse(vector);
+        Mockito.verifyNoMoreInteractions(singleSentenceResultListener);
+    }
+
+    public void testInferenceSentences_whenGetModelException_thenFailure() {
```

Comment:
we need to test scenario when we're retrying 1-2 times. I see scenario when first request has failed with error that isn't retryable, this isn't a full coverage

Reply:
@martin-gaievski thanks for pointing this out.

I'm wondering though, if we should make this retryable at all. Let me elaborate:

In my understanding, inference requests are retried because they tend to fail more often than regular operations in OpenSearch. I don't know the history and complete reasoning behind this, so I speculate it has to do with the fact that the inference is done natively and that many things can go wrong there.

With my change, if fetching the model information fails (`mlClient.getModel(modelId, ...`), there is no retry. Model information is fetched the first time inference is requested with a particular model. After that, the result is cached and the method behaves exactly as before the PR.

So my argument is: should we really add a retry logic to this relatively simple operation? If `getModel` fails, it is most likely to fail again, so retrying wouldn't make sense. If so, one could argue that all operations in OpenSearch should be wrapped in a retry logic.

Reply:
@br3no I think Martin was talking about adding more test method to increase the test coverage, e.g. add a new test method `testInferenceSentences_whenGetModelNodeDisconnectedException_thenRetryToSuccess`, with this method you only need to change the 
```
RuntimeException exception = new RuntimeException("Bam!");
setupMocksForTextEmbeddingModelAsymmetryCheck(exception);
```
to
```
NodeDisconnectedException exception = new NodeDisconnectedException("node disconnected!");
setupMocksForTextEmbeddingModelAsymmetryCheck(exception);
```
in this method, and by that the new method should cover the retry part.

Reply:
@zane-neo if I replace the exception, the behavior will not change, i.e. there will not be a successful retry. 

This is because the `checkModelAsymmetryAndThenPredict` method in `MLCommonsClientAccessor` will not retry `mlClient.getModel` if there is a failure.

The reasoning was that if there is an exception in `mlClient.getModel`, which is a very simple operation (it's just fetching the model information), then most likely retrying it will lead to the same failure. If we think this operation should be retried, then, I argued above, all operations in OpenSearch would be equally good candidates to be wrapped in a retry logic.

I'm happy to implement the retry in this operation, but I don't believe it's worth it. It will introduce more complexity to an already quite intricate code path, making it less readable. At the same time I believe it will almost never really be useful, as a failing `mlClient.getModel` will almost always fail again if repeated.

Reply:
Agreed, it's fine to not retry the `client.getModel` since it's essentially different with `client.predict`, for predict request, ml-commons has internal routing logic to dispatch the request to different nodes and it's possible that the node disconnected from cluster right after received the request, then retry is useful as next predict request will be dispatched to a different node in round robin approach. But for `client.getMode` it internally uses the `client.get` operation which is more deterministic, e.g. user might specify to read the result from primary shard or read from the only one replica, retry won't help in these cases. And it's a valid argue that in OpenSearch not all the operations are wrapped with retry logic.

Reply:
@zane-neo @br3no please help me to understand the flow for reading model config. Is there a chance that this call is remote, meaning going outside the opensearch cluster if model deployed remotely and connected via ml-commons remote connector? 

If that's not the case then I'm fine with logic without retry, it's a transport request within single OS domain. In this case where the mode configuration is stored, is it in ml-commons memory or cluster metadata?

If it can be a remote call then we need a retry. Doesn't matter how simple operation is, there can be a transient network issue or anything of this sort. In this case it's different from most of OS APIs, those are primality case 1, transport requests inside cluster.

Reply:
No case is to fetching model metadata from remote services, cause they are all stored in OS cluster.